go-update/src/main/java/org/reactome/release/goupdate/DuplicateCleaner.java

```diff
@@ -0,0 +1,215 @@
+			if (refCounts.size() > 0)
+			{
+				// for each DB ID in referrer counts
+				for (Long dbId : refCounts.keySet())
+				{
+					// if there are referrers, we need to report them.
+					if (refCounts.get(dbId) > 0)
+					{
+						// update the counter of all instances with significant referrers.
+						instancesWithSignificantReferrers ++;
+						// log info about the referrers 
+						logReferrers(dbId, refCounts);
+					}
+					// if there are NO referrers, add the db_id to the list of DB IDs without referrers.
+					else
+					{
+						logger.debug("DB ID {} for accession {} has no significant referrers, and will be deleted.", dbId, accession);
+						dbIDsWithNoReferrers.add(dbId);
+					}
+				}
+			}
+			// update list of things to delete with IDs that have no referrers.
+			dbIDsToDelete.addAll(dbIDsWithNoReferrers);
+			// If no DB IDs have significant referrers, we should delete ALL DB_IDs for the accession, except for the newest one.
+			if (dbIDsWithNoReferrers.size() == refCounts.keySet().size())
+			{
+				GKInstance newestInstance = null;
+				// Need to try all classes since we don't know which class a GO Accession might be. Note: this.goClasses also contains "Compartment", but here "Compartment" is not used.
+				// It might be safe to use this.goClasses here, but since this bit of code did not use Compartment, we are not going to use this.goClasses here, for now.
+				for (String reactomeClass : Arrays.asList(ReactomeJavaConstants.GO_BiologicalProcess, ReactomeJavaConstants.GO_MolecularFunction, ReactomeJavaConstants.GO_CellularComponent))
+				{
+					@SuppressWarnings("unchecked")
+					Collection<GKInstance> instances = (Collection<GKInstance>) adaptor.fetchInstanceByAttribute(reactomeClass, ReactomeJavaConstants.accession, "=", accession);
+					// For each instance for this accession, try to find the newest one so that it won't be deleted.
+					if (instances != null && instances.size() > 0)
+					{
+						// Add all the DB_IDs to the list to delete. Later, only the DB_ID of the NEWEST instance will be removed from this list, and all other older instances will be deleted.
+						dbIDsToDelete.addAll(instances.stream().map(i -> i.getDBID()).collect(Collectors.toSet()));
+						newestInstance = findNewestInstance(instances);
+					}
+				}
+				logger.debug("Newest instance for accession {} is {} - this instance will NOT be deleted.", accession, newestInstance.toString());
+				// REMOVE the ID of the newest instance from the list of things to delete (it would have been automatically added if the "newest" instance
+				// had 0 referrers).
+				dbIDsToDelete.remove(newestInstance.getDBID());
+			}
+		}
+		logger.info("\n\nSummary:\nTotal number of duplicated accessions: {} \n"
+				+ "Number of instances with significant (non-GO Term) referrers: {}\n", countOfDuplicatesAccessions, instancesWithSignificantReferrers);
```

Comment:
The `countOfDuplicatesAccessions` variable can be removed and replaced directly with the query `duplicates.keySet().size()`.

Reply:
I'm not sure, but it might be possible (or it _was_ possible, at one time) for `duplicates` to be null, in which case `duplicates.keySet()...` would throw a `NullPointerException`.